tls/s2n_client_extensions.c

```diff
@@ -226,6 +226,64 @@ int s2n_client_extensions_recv(struct s2n_connection *conn, struct s2n_blob *ext
     return 0;
 }
 
+uint32_t s2n_client_hello_get_extension_length(struct s2n_client_hello *ch, s2n_tls_extension_type extension_type)
+{
+    notnull_check(ch);
+    notnull_check(ch->extensions.data);
+
+    struct s2n_stuffer in;
+    GUARD(s2n_stuffer_init(&in, &ch->extensions));
+    GUARD(s2n_stuffer_write(&in, &ch->extensions));
+
+    while (s2n_stuffer_data_available(&in)) {
+        uint16_t ext_type, ext_size;
+
+        GUARD(s2n_stuffer_read_uint16(&in, &ext_type));
+        GUARD(s2n_stuffer_read_uint16(&in, &ext_size));
+
+        if (extension_type == ext_type) {
+            return ext_size;
+        }
+
+        /* Move the pointer ahead */
+        s2n_stuffer_raw_read(&in, ext_size);
+    }
+
+    return 0;
+}
+
+uint32_t s2n_client_hello_get_extension_by_id(struct s2n_client_hello *ch, s2n_tls_extension_type extension_type, uint8_t *out, uint32_t max_length)
```

Comment:
I thought about it while I was writing these functions and I chose to do it this way for two reasons:
* I was not sure how often these calls will be used and give that the extensions are not that many I felt it should be pretty fast to parse.
* Saving extensions map will add additional memory overhead (which might not be too significant).

Let me know if you feel strongly about moving extensions to a map.

Reply:
trading memory for speed makes sense, if we can keep the footprint small, for example storing data pointers into the client hello rather than copying the bytes.  Another advantage this gives us is the ability to detect duplicate extensions with the same ID, which should result in the server aborting the handshake.

Given the set of registered extesions (see https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml) exposes only 27 today, a sorted array would outperform a hash/map on both speed and memory.

Reply:
Working on using array here, will update soon.
I will work on aborting server handshake with duplicate extensions later along with updating:
https://github.com/awslabs/s2n/blob/master/tls/s2n_client_hello.c#L50-L107 api calls to return int instead of uint32_t since notnull_check() will return -1 and cause overflow.